monitoring/ping.go

```diff
@@ -0,0 +1,314 @@
+	latencyQuantile = 95.0
+)
+
+// pingChecker is a checker that verifies that ping times (RTT) between nodes in
+// the cluster are within a predefined threshold
+type pingChecker struct {
+	self           serf.Member
+	serfClient     serf.RPCClient
+	serfRPCAddr    string
+	serfMemberName string
+	latencyStats   ttlmap.TTLMap
+	mux            sync.Mutex
+	logger         log.Entry
+}
+
+// NewPingChecker returns a checker that verifies accessibility of nodes in the cluster by exchanging ping requests
+func NewPingChecker(serfRPCAddr string, serfMemberName string) (c health.Checker, err error) {
+	latencyTTLMap, err := ttlmap.New(int(statsTTLPeriod.Seconds())) //FIXME: why is number of seconds used as capacity?
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	logger := log.WithFields(log.Fields{trace.Component: "ping"})
+	logger.Debugf("using Serf IP: %v", serfRPCAddr)
+	logger.Debugf("using Serf Name: %v", serfMemberName)
+
+	client, err := newSerfClient(serfRPCAddr)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// retrieve other nodes using Serf members
+	nodes, err := client.Members()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	// finding what is the current node
+	var self serf.Member
+	for _, node := range nodes {
+		if node.Name == serfMemberName {
+			self = node
+			break // self node found, breaking out of the for loop
+		}
+	}
+	if self.Name == "" {
+		return nil, trace.NotFound("failed to find Serf member with name %s", serfMemberName)
+	}
+
+	return &pingChecker{
+		self:           self,
+		serfClient:     *client,
+		serfRPCAddr:    serfRPCAddr,
+		serfMemberName: serfMemberName,
+		latencyStats:   *latencyTTLMap,
+		logger:         *logger,
+	}, nil
+}
+
+func newSerfClient(serfRPCAddr string) (client *serf.RPCClient, err error) {
+	// fetch serf config and instantiate client
+	clientConfig := serf.Config{
+		Addr: serfRPCAddr,
+	}
+	client, err = serf.ClientFromConfig(&clientConfig)
+	if err != nil {
+		return client, trace.Wrap(err)
+	}
+	return client, nil
+}
+
+// Name returns the checker name
+// Implements health.Checker
+func (c *pingChecker) Name() string {
+	return pingCheckerID
+}
+
+// Check verifies that all nodes' ping with Master Nodes is lower than the
+// desired threshold
+// Implements health.Checker
+func (c *pingChecker) Check(ctx context.Context, r health.Reporter) {
+	// FIXME: #1 RTTThreshold will become configurable in future
+	// FIXME: #2 Send RTTThreshold value to metrics
+
+	err := c.check(ctx, r)
+
+	if err != nil {
+		c.setProbeStatus(ctx, r, err, pb.Probe_Failed)
+		return
+	}
+	c.setProbeStatus(ctx, r, nil, pb.Probe_Running)
+
+	return
+}
+
+// check runs the actual system status verification code and returns an error
+// in case issues arise in the process
+func (c *pingChecker) check(ctx context.Context, r health.Reporter) (err error) {
+
+	client := &c.serfClient
+	// check if client connection closed and reopen it
+	if client.IsClosed() {
```

Comment:
This comes from a comment @a-palchikov added about a case in which the Serf connection is unstable (or the cluster is restarted) and the connection gets closed.
Thus the check plus the client reopening.

Reply:
@eldios @a-palchikov Isn't serf client self-reconnecting though? So if connection was closed, it will detect it and reconnect on the `Members` call below? https://github.com/gravitational/satellite/blob/master/agent/serf.go#L56